zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -1276,13 +2068,672 @@ impl<F: Field> PiCircuitConfig<F> {
+                    offset,
+                    || Value::known(F::from((*rlp_byte) as u64).invert().unwrap_or(F::zero())),
+                )
+                .unwrap();
+            region
+                .assign_advice(
+                    || "q_blk_hdr_rlc_acc",
+                    self.blockhash_cols.q_blk_hdr_rlc_acc,
+                    offset,
+                    || Value::known(F::from(q_blk_hdr_rlc_acc[offset] as u64)),
+                )
+                .unwrap();
+            region
+                .assign_advice(
+                    || "blk_hdr_rlc_acc",
+                    self.blockhash_cols.blk_hdr_rlc_acc,
+                    offset,
+                    || blk_hdr_rlc_acc[offset],
+                )
+                .unwrap();
+            region
+                .assign_advice(
+                    || "blk_hdr_is_leading_zero",
+                    self.blockhash_cols.blk_hdr_is_leading_zero,
+                    offset,
+                    || Value::known(F::from(leading_zeros[offset] as u64)),
+                )
+                .unwrap();
+
+            self.blockhash_cols
+                .q_blk_hdr_rlp
+                .enable(region, offset)
+                .unwrap();
+        }
+
+        // Gets rid of CellNotAssigned occuring in the last row
+        for fixed_col in [
+            self.blockhash_cols.q_number,
+            self.blockhash_cols.q_var_field_256,
+            self.blockhash_cols.q_reconstruct,
+        ] {
+            region
+                .assign_fixed(
+                    || "fixed column last row",
+                    fixed_col,
+                    BLOCKHASH_TOTAL_ROWS,
+                    || Value::known(F::zero()),
+                )
+                .unwrap();
+        }
+
+        for advice_col in [
+            self.blockhash_cols.blk_hdr_rlc_acc,
+            self.blockhash_cols.blk_hdr_rlp,
+            self.blockhash_cols.blk_hdr_reconstruct_value,
+        ] {
+            region
+                .assign_advice(
+                    || "advice column last row",
+                    advice_col,
+                    BLOCKHASH_TOTAL_ROWS,
+                    || Value::known(F::zero()),
+                )
+                .unwrap();
+        }
+
+        // Calculate reconstructed values
+        let mut reconstructed_values: Vec<Vec<Value<F>>> = vec![];
+        let randomness = challenges.evm_word();
+        for value in [
+            public_data.parent_hash.as_fixed_bytes().iter(),
+            public_data.beneficiary.as_fixed_bytes().iter(),
+            public_data.state_root.as_fixed_bytes().iter(),
+            public_data.transactions_root.as_fixed_bytes().iter(),
+            public_data.receipts_root.as_fixed_bytes().iter(),
+            public_data
+                .block_constants
+                .number
+                .as_u64()
+                .to_be_bytes()
+                .iter(),
+            public_data.block_constants.gas_limit.to_be_bytes().iter(),
```

Comment:
gas_limit, number, gas_used and difficult is in number encoding, do we still need to rlc them??

Reply:
gas limit, gas used and difficulty are 256bit. I had two options for checking them:
1. reconstruct their value from their byte representation and split them to lo/hi of 128bits each. This ends up using 2 lookup operations
2. use RLC to reduce them to a finite field value. This uses 1 lookup operation

All these circuits will be executed 257 times (for previous hashes too), so we save 257 lookups. 

The number field could be checked as literal value instead of RLC as it is only 8 bytes. I used RLC to keep a uniform approach.

Reply:
Seems like the hi/lo split on the PSE repo is making great progress, and I didn't wan to work against something that was only half speced out, but the split in hi/lo will very likely be the way forward because that's also how the data will need to be stored in the lookup tables that are accessed by different circuits. Just using rlc for the keccak input data checking and perhaps some internal data checking of dynamically sized data where necessary.

So rlc is good for now, but keep in mind that'll very likely have to change to at least having the lo/hi split parts as well.

> All these circuits will be executed 257 times (for previous hashes too), so we save 257 lookups.

Most important is just the number of lookups per row. But even with two 128bit parts it would still be a single lookup no? Doing a lookup per part would not work because that only verifies that each of those two parts are in the lookup table separately, but we need to make sure we have both of them at the expected location together.

Reply:
Just forget why we need do that for previous 256 hashes? I remember prev hashes are located in L1 and are part of anchor tx, so they are verified in anchor?? @Brechtpd